### PR TITLE
silence unused warning on MSVC

### PR DESCRIPTION
**Pull request by maxbachmann**
@scoder @da-woods This is an alternative implementation of #4058 since it does not appear like there is actually a way to suppress a warning a single time on MSVC.

- Would a change like this be fine, since it requires more code than the previous solution?
- Would we still keep the `CYTHON_UNUSED` macro for function arguments + local variables. `CYTHON_UNUSER_VAR` already silences the warning. So the only reason to keep `CYTHON_UNUSED` in these places would be readability.

After this change there is still the warning about unused functions. However the biggest part of the warnings is about unuser parameters/variables anyways.

### CODE REVIEW

**Comment by scoder** [comment]
Big change, touching a lot of code including pending PRs/branches.
But I think it's better than what we currently have, more universal.

The `CYTHON_UNUSED` annotations can go away, I think. It's declarative enough when the function starts with `CYTHON_UNUSED_VAR(…)`.

Please take care that Cython can still generate C89 code, mostly for MSVC. I think I spotted a couple of "code before declaration" cases (`(void)…` is code…).

File: Cython/Compiler/Nodes.py [hunk 1]
```diff
@@ -2721,6 +2724,9 @@ def generate_function_header(self, code, with_pymethdef, with_opt_args=1, with_d
             code.globalstate.parts['module_declarations'].putln(
                 "%s%s%s; /* proto*/" % (storage_class, modifiers, header))
         code.putln("%s%s%s {" % (storage_class, modifiers, header))
+        # silence unused warnings when CYTHON_UNUSED does not work (e.g. on msvc)
+        for arg in unused_args:
+            code.putln("CYTHON_UNUSED_VAR(%s);" % arg)
```
Comment (maxbachmann) [inline Cython/Compiler/Nodes.py:2729]: How can I fix this for C89? I currently simply generate them directly behind the opening brackets. However there could be variable declarations afterwards.

File: Cython/Compiler/Code.py [hunk 2]
```diff
@@ -2068,15 +2068,15 @@ def put_var_declaration(self, entry, storage_class="",
...
-            self.put('CYTHON_UNUSED ')
         self.put(entry.type.declaration_code(
             entry.cname, dll_linkage=dll_linkage))
         if entry.init is not None:
             self.put_safe(" = %s" % entry.type.literal_code(entry.init))
         elif entry.type.is_pyobject:
             self.put(" = NULL")
         self.putln(";")
+        if not entry.cf_used:
+            self.putln("CYTHON_UNUSED_VAR(%s);" % entry.cname)
```
Comment (maxbachmann) [inline Cython/Compiler/Code.py:2079]: wrong order for c89

File: Cython/Compiler/Code.py [hunk 3]
```diff
@@ -2089,14 +2089,14 @@ def put_temp_declarations(self, func_context):
...
-            self.putln("%sint %s = 0;" % (unused, Naming.lineno_cname))
-            self.putln("%sconst char *%s = NULL;" % (unused, Naming.filename_cname))
-            self.putln("%sint %s = 0;" % (unused, Naming.clineno_cname))
+            self.putln("int %s = 0;" % Naming.lineno_cname)
+            self.putln("const char *%s = NULL;" % Naming.filename_cname)
+            self.putln("int %s = 0;" % Naming.clineno_cname)
+            if not self.funcstate.uses_error_indicator:
+                self.putln("CYTHON_UNUSED_VAR(%s);" % Naming.lineno_cname)
+                self.putln("CYTHON_UNUSED_VAR(%s);" % Naming.filename_cname)
+                self.putln("CYTHON_UNUSED_VAR(%s);" % Naming.clineno_cname)
```
Comment (maxbachmann) [inline Cython/Compiler/Code.py:2099]: probably same

Comment (da-woods) [reply Cython/Compiler/Code.py:2099]: You might use `code.insertion_point` to create a sub-writer for all the declarations (thus ensuring they end up first)

**Comment by mattip** [comment]
> a way to suppress a warning a single time on MSVC

There is [`push` and `pop`](https://docs.microsoft.com/en-us/cpp/preprocessor/warning?view=msvc-170)

**Comment by maxbachmann** [comment]
> There is [push and pop](https://docs.microsoft.com/en-us/cpp/preprocessor/warning?view=msvc-170)

yes but in this context this can not be used:
```
void func(CYTHON_UNUSED int var) {}
```

**Comment by mattip** [comment]
Ahh, thanks for the explanation.

**Pull request closed without merging**